### ISSUE TITLE
Fetch module import failed, trying to import socket.io-client in the browser.

### ISSUE THREAD

**Issue by OSG-Proj** [socketio/socket.io#3594]
**Note**: for support questions, please use one of these channels: [stackoverflow](http://stackoverflow.com/questions/tagged/socket.io) or [slack](https://socketio.slack.com)

For bug reports and feature requests for the **Swift client**, please open an issue [there](https://github.com/socketio/socket.io-client-swift).

For bug reports and feature requests for the **Java client**, please open an issue [there](https://github.com/socketio/socket.io-client-java).

### You want to:

* [x] report a *bug*
* [ ] request a *feature*

### Current behaviour

*What is actually happening?*
I am using the es6 import syntax, but it's not working, while intellisense works for it in vs code
### Steps to reproduce (if the current behaviour is a bug)
1. import { io } from 'socket.io-client'
2. Put JS file on html page
3. Run express http server
4. Chrome says 'Fetch module import failed'
**Note**: the best way (and by that we mean **the only way**) to get a quick answer is to provide a failing test case by forking the following [fiddle](https://github.com/socketio/socket.io-fiddle).

### Expected behaviour

*What is expected?
For this to work, as documented on the website
### Setup
- OS: windows 10
- browser: cheome
- socket.io version: 2.3.0

### Other information (e.g. stacktraces, related issues, suggestions how to fix)
Rewrite the socket-io client to support es6 modules, using only export defaults in socket.io-client, as nodejs is used as server side.

**Comment by iakshat** [comment]
If I am correct what you are trying to do is use this line:

`import { io } from 'socket.io-client'`

in the HTML file that you serve through the express server.
Try using this instead:

`<script src="/socket.io/socket.io.js"></script>`

The browser will not have access to the node modules you've installed for the express server and that's why it dosen't work.

**Comment by OSG-Proj** [comment]
Yeah but I'm trying to use es6 import method

**Comment by iakshat** [comment]
Okay then try using this:

`<script type="module">
import {io} from '<link to socket.io library>'
</script>`

It seems import statements work only in module type script tags

**Comment by jack0088** [comment]
@iakshat I have the exact same issue! I want to use es6 import method in the browser.

- ExpressJS serves the socket.io client library automatically at `/socket.io/socket.io.js`.
- My http server responds with this HTML: `<!doctype html><html><head><script src="/hello.js" type="module"></script></head><body></body></html>` (Note, that it's a type="module"!)
- *hello.js* imports then socket.io like this: `import {io} from "/socket.io/socket.io.js";` and that produces an error: `SyntaxError: Importing binding name 'io' is not found.` So your suggestion sadly doesn't work!
If I try to simply import it like this: `import "/socket.io/socket.io.js";` it also fails with the error `TypeError: undefined is not an object (evaluating 't.io=e()')`. I think in this case it can't access the window object or something - as `t.io` would be an assignment to `this.io`, which in turn refers to `window.io`.

I would really like to NOT put `<script src="/socket.io/socket.io.js"></script>` into my HTML. Do you have any suggestions, please?

**Comment by iakshat** [comment]
@jack0088  thanks for pointing out this mistake.

After some attempts, I found out that the import syntax is fine and it works for a simple module with export(on which my previous answer was based on) but the script in "/socket.io/socket.io.js" is a build of the library and does not have io as export. On the backend with node, you can import the client library as

`import io from 'socket.io-client';`

and after checking out that file I don't think it's easy to import it directly. Using the build as script-src turns out to be the only option.

**Comment by OSG-Proj** [comment]
Actually I found out that the module is a commonjs, but using the es6 import will work using webpack. Using import io from 'socket.io-client' works because webpack detects the module by the folder name is node_modules and looks for an index.js, same way how express is imported. It's strange how es6 imports work with commonjs

**Comment by jmonster** [comment]
Running into this same problem while playing with _lit-element_; I found 2 syntaxes that get past this error but yield another.

```
var promise = import('socket.io-client');
```

yields

```
Unhandled Promise Rejection: ReferenceError: Can't find variable: require
```

**Comment by darrachequesne** [comment]
For future readers:

The `io` method is bound to the global scope in the standalone build:

```html
<script src="/socket.io/socket.io.js"></script>
<script>
  const socket = io();
</script>
```

An ESM bundle is also available since version [4.3.0](/blog/socket-io-4-3-0/):

```html
<script type="module">
  import { io } from "https://cdn.socket.io/4.4.1/socket.io.esm.min.js";

  const socket = io();
</script>
```

With an [import map](https://caniuse.com/import-maps):

```html
<script type="importmap">
  {
    "imports": {
      "socket.io-client": "https://cdn.socket.io/4.4.1/socket.io.esm.min.js"
    }
  }
</script>

<script type="module">
  import { io } from "socket.io-client";

  const socket = io();
</script>
```

Else, in all other cases (with some build tools, in Node.js or React Native), it can be imported from the `socket.io-client` package:

```js
// ES modules
import { io } from "socket.io-client";

// CommonJS
const { io } = require("socket.io-client");
```

Reference: https://socket.io/docs/v4/client-api/

**Comment by coloralp** [comment]
> Okay then try using this:
> 
> `<script type="module"> import {io} from '<link to socket.io library>' </script>`
> 
> It seems import statements work only in module type script tags

Thanks a lot
